### PR TITLE
opt: add implicit SELECT FOR UPDATE to initial scan of DELETE

### LINKED ISSUE #50181: sql/opt: add implicit SELECT FOR UPDATE support for DELETE statements
Address this TODO:

https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/opt/exec/execbuilder/mutation.go#L991-L999

Similar to https://github.com/cockroachdb/cockroach/issues/50180, but likely lower priority because sustained contending DELETEs don't seem to be as common, probably because once a row is deleted, it stays deleted and can't be deleted again.

Jira issue: CRDB-4147

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit makes it so that we apply implicit SELECT FOR UPDATE locking behavior to the initial scan of the DELETE operation in some cases. Namely, we do so when the input to the DELETE is either a scan or an index join on top of a scan (with any number of renders on top) - these conditions mean that all filters were pushed down into the scan, so we won't lock any unnecessary rows. I think it's only possible to have at most one render expression on top of the scan, but I chose to be defensive and allowed nested renders too. In such form the conditions are exactly the same as we use for adding SFU to UPDATEs, so the same function is reused. Existing `enable_implicit_select_for_update` session variable is consulted.

Fixes: #50181.

Release note (sql change): DELETE statements now acquire locks using the FOR UPDATE locking mode during their initial row scan in some comes, which improves performance for contended workloads. This behavior is configurable using the `enable_implicit_select_for_update` session variable.

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137069)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODiC8hqE1AUrP6PkFTg:-ODiC8hqE1AUrP6PkFTh:b-4z10og) ([raw file](https://github.com/cockroachdb/cockroach/blob/8935b1f19c5a99efc83227bd4410f1ac452daba4/pkg/sql/opt/exec/execbuilder/mutation.go#L1238)):*
> ```Go
> // DELETE statement. If the builder should lock the initial row scan, it returns
> // the TableID of the scan, otherwise it returns 0.
> func (b *Builder) shouldApplyImplicitLockingToDeleteInput(del *memo.DeleteExpr) opt.TableID {
> ```

Given that the code is an exact copy of the UPDATE case, should we share the same helper between two mutations? I'm currently leaning towards keeping them separate since the amount of duplication is very low and it seems cleaner this way.

**Review by michae2 — COMMENTED** [review]
Nice! That was easier than I realized it would be!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-:-ODwSnz_CGDblSkIu7v9:bxkus3w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODiC8hqE1AUrP6PkFTg:-ODwS89h4_XdSOLzwBLV:bto22el) ([raw file](https://github.com/cockroachdb/cockroach/blob/8935b1f19c5a99efc83227bd4410f1ac452daba4/pkg/sql/opt/exec/execbuilder/mutation.go#L1238)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Given that the code is an exact copy of the UPDATE case, should we share the same helper between two mutations? I'm currently leaning towards keeping them separate since the amount of duplication is very low and it seems cleaner this way.
</blockquote></details>

Coincidentally, @mgartner just changed this function for updates during the collab session today, so actually maybe it would be good to share the helper function?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODwSVdTCxAXQi1I7W2Z:-ODwSVdTCxAXQi1I7W2_:bxf16dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a0f4808c5066db1db213f391a35306bd51335/pkg/sql/opt/exec/execbuilder/mutation.go#L1134)):*
> ```Go
> 	mutExpr memo.RelExpr,
> ) (opt.TableID, error) {
> 	if !b.evalCtx.SessionData().ImplicitSelectForUpdate {
> ```

Does moving the call here also affect `INSERT INTO SELECT FROM` statements? I don't think we would want to affect those.

**Review by mgartner — APPROVED** [review]
:lgtm:

> Reviewed 14 of 14 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-:-ODwoIK99CK_PR6v-Ych:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODiC8hqE1AUrP6PkFTg:-ODwnSal0uJk2-cNR9I-:b-nmo6an) ([raw file](https://github.com/cockroachdb/cockroach/blob/8935b1f19c5a99efc83227bd4410f1ac452daba4/pkg/sql/opt/exec/execbuilder/mutation.go#L1238)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Coincidentally, @mgartner just changed this function for updates during the collab session today, so actually maybe it would be good to share the helper function?
</blockquote></details>

Either way works for me. If you want to combine, it shouldn't be too much work because the function can take a `memo.RelExpr` and you can use `.Child(0).(RelExpr)` to access the input.

I'll hold off on merging my PR until this one merges.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 347 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODwo1Z94crVjv6knWFi:-ODwo1Z94crVjv6knWFj:b6picd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a0f4808c5066db1db213f391a35306bd51335/pkg/sql/opt/exec/execbuilder/testdata/delete#L347)):*
> ```
>               spans: /1-/1000 /2001-/3000
>               locking strength: for update
> ```

nit: Some tests cases specifically for testing the different cases of implicit FOR UPDATE locking might be nice. Up to you.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137069)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODiC8hqE1AUrP6PkFTg:-ODws9L561e__jgLxymG:b5vo7wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8935b1f19c5a99efc83227bd4410f1ac452daba4/pkg/sql/opt/exec/execbuilder/mutation.go#L1238)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Either way works for me. If you want to combine, it shouldn't be too much work because the function can take a `memo.RelExpr` and you can use `.Child(0).(RelExpr)` to access the input.

I'll hold off on merging my PR until this one merges.
</blockquote></details>

Ack, combined them into one.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODwSVdTCxAXQi1I7W2Z:-ODwqdxcD88B_5cuFjOg:b-j71dqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a0f4808c5066db1db213f391a35306bd51335/pkg/sql/opt/exec/execbuilder/mutation.go#L1134)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Does moving the call here also affect `INSERT INTO SELECT FROM` statements? I don't think we would want to affect those.
</blockquote></details>

We get exactly the same behavior for INSERTs as before - it has the same return values as this clause, so this change should be a no-op for INSERTs.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/delete` line 347 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-ODwo1Z94crVjv6knWFi:-ODwsSEs4KXituiFFRxC:bkvh8lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a0f4808c5066db1db213f391a35306bd51335/pkg/sql/opt/exec/execbuilder/testdata/delete#L347)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Some tests cases specifically for testing the different cases of implicit FOR UPDATE locking might be nice. Up to you.
</blockquote></details>

I think existing tests have decent variability covering any simple cases that I would add already.

**Review by michae2 — APPROVED** [review]
:lgtm: Awesome.

> Reviewed 13 of 14 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137069#-:-ODxBoJkDUG8niX8vl0i:bswp6ey)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Comment by yuzefovich** [comment]
TFTRs!

bors r+

**Pull request merged**
